src/main/java/io/gravitee/el/spel/CachedExpression.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.el.spel;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import org.springframework.expression.Expression;
+import org.springframework.expression.common.CompositeStringExpression;
+import org.springframework.expression.spel.SpelNode;
+import org.springframework.expression.spel.ast.CompoundExpression;
+import org.springframework.expression.spel.ast.PropertyOrFieldReference;
+import org.springframework.expression.spel.ast.VariableReference;
+import org.springframework.expression.spel.standard.SpelExpression;
+
+/**
+ * Allows to cache a given {@link Expression} for later reuse.
+ * It also gives access to the list of variables accessed by the {@link Expression}.
+ *
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class CachedExpression {
+
+    private final Expression expression;
+    private Set<String> variables;
+
+    public CachedExpression(final Expression expression) {
+        this.expression = expression;
+        computeVariables(expression);
+    }
+
+    private void computeVariables(Expression expression) {
+        if (expression instanceof SpelExpression) {
+            computeVariables(((SpelExpression) expression).getAST());
+        } else if (expression instanceof CompositeStringExpression) {
+            computeVariables((CompositeStringExpression) expression);
+        }
+    }
+
+    private void computeVariables(CompositeStringExpression expression) {
+        for (Expression e : expression.getExpressions()) {
+            computeVariables(e);
+        }
+    }
+
+    private void computeVariables(CompoundExpression expression) {
```

Comment:
maybe it worth a bit of java doc to ease the reading of this method ?

Reply:
do you think it will evolve to handle another specialization of Expression ? Does it worth to have the algo in dedicated class for `CompositeStringExpression` and `CompoundExpression` ?

Reply:
It seems we currently cover all known usages we have with EL. 
Maybe in the futur we will decide to add some engineering touch if we found new cases to cover 🙂